File: lib/__builtins__/buffer.py
#!/usr/bin/env python

"""
Buffer object.

Copyright (C) 2015, 2016, 2017, 2021 Paul Boddie <paul@boddie.org.uk>

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.
"""

from native import list_init, list_append, list_concat, buffer_str

class buffer:

    "A buffer, used to build strings."

    def __init__(self, args=None, size=0):

        "Initialise a buffer from the given 'args' or the given 'size'."

        if args is not None:
            n = len(args)
        elif isinstance(size, integer):
            n = size
        else:
            raise ValueError(size)

        self.__data__ = list_init(n)

        # Append all arguments to the buffer.

        if args:
            for arg in args:
                self.append(arg)

    def append(self, s):

        """
        Append 's' to the buffer, concatenating buffers and adding other objects
        in string form.
        """

        if isinstance(s, buffer):
            list_concat(self, s.__data__)
        elif isinstance(s, string):
            list_append(self, s)
        else:
            list_append(self, str(s))

    def __str__(self):

        "Return a string representation."

        return buffer_str(self.__data__)

    def __repr__(self):

        "Return a program representation."

        return buffer(["buffer([", repr(str(self)), "])"])

# vim: tabstop=4 expandtab shiftwidth=4

File: templates/native/buffer.c
/* Native functions for buffer operations.

Copyright (C) 2016, 2017, 2021 Paul Boddie <paul@boddie.org.uk>

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <string.h> /* strcmp, memcpy */
#include "native/common.h"
#include "types.h"
#include "exceptions.h"
#include "ops.h"
#include "progconsts.h"
#include "progops.h"
#include "progtypes.h"
#include "main.h"

__attr __fn_native_buffer_buffer_str(__attr __self, __attr _data)
{
    /* _data interpreted as buffer.__data__ */
    __fragment *data = _data.seqvalue;
    __int size = 0, i, j, n;
    char *s;
    __attr o;

    /* Calculate the size of the string. */
    for (i = 0; i < data->size; i++)
        size += __TOINT(__load_via_object(__VALUE(data->attrs[i]), __size__));

    /* Reserve space for a new string. */
    s = (char *) __ALLOCATE(size + 1, sizeof(char));

    /* Build a single string from the buffer contents. */
    for (i = 0, j = 0; i < data->size; i++)
    {
        o = __load_via_object(__VALUE(data->attrs[i]), __data__);
        n = __TOINT(__load_via_object(__VALUE(data->attrs[i]), __size__));
        memcpy(s + j, o.strvalue, n); /* does not null terminate but final byte should be zero */
        j += n;
    }

    /* Return a new string. */
    return __new_str(s, size);
}

/* Module initialisation. */

void __main_native_buffer()
{
}
